cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ssa0/SsaInternals.qll

```diff
@@ -87,12 +87,35 @@ class DefImpl extends DefOrUseImpl, TDefImpl {
   }
 
   predicate isCertain() { isDef(true, _, address, _, _, _) }
+
+  /**
+   * Holds if this definition not having a use does not imply that it
+   * will be removed from the SSA def-use relation.
+   */
+  predicate cannotBePruned() { none() }
 }
 
-class UseImpl extends DefOrUseImpl, TUseImpl {
+private module IteratorDefUse {
+  import semmle.code.cpp.models.interfaces.Iterator as Interfaces
+  import semmle.code.cpp.models.implementations.Iterator as Iterator
+  import semmle.code.cpp.models.implementations.StdContainer as StdContainer
+
+  class IteratorDef extends DirectDef {
+    IteratorDef() {
+      exists(Type t | t = this.getBaseSourceVariable().getType().getUnspecifiedType() |
+        t instanceof Iterator::Iterator and
+        not t instanceof PointerOrReferenceType
```

Comment:
Why is it needed to exclude `PointerOrReferenceType` here? Can an iterator be a pointer or reference type?

Reply:
Yep: https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/models/implementations/Iterator.qll#L48. I'd really like to remove that class actually. It exists because it was a quick fix to a customer issue in AST dataflow a few years ago.

Reply:
😢